### ISSUE TITLE
Rustc keeps suggesting to increase type_length_limit instead of giving the final value

### ISSUE THREAD

**Issue by NotWearingPants** [rust-lang/rust#72755]
I tried compiling this code:

```rust
fn main() {
	let x = vec!(1, 2, 3);

	let y: std::vec::Vec<u32> = x.iter()
		.map(|byte| *byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.map(|byte| byte)
		.collect();

	println!("{}", y[0]);
}
```
(don't ask why... lol)

I expected it to compile and print 1, because I just have some maps which don't affect the vector.

Instead, the compiler printed:
```
error: reached the type-length limit while instantiating `<std::iter::Map<std::iter::Map<s...osure@src\main.rs:13:8: 13:19]]>`
    |
    = note: consider adding a `#![type_length_limit="1376294"]` attribute to your crate

error: aborting due to previous error
```

I have no idea why it fails, but I do as it says and add the `type_length_limit="1376294"` line to the top of the file.

Recompiling, it gives the same output but now with `type_length_limit="2752543"`.

After using the new value and recompiling, it now wants 5505048.

Repeating the same process, I get 11010065, then 22020106, then 44040195, etc.
Basically it keeps almost doubling the value each time, and my code still doesn't compile for some reason.

The compiler should either give me the final value that would work, or fail with a better error message if no value would work, or better - succeed in compiling this simple program that would simply work in other programming languages.

### Meta
`rustc --version --verbose`:
```
rustc 1.43.1 (8d69840ab 2020-05-04)
binary: rustc
commit-hash: 8d69840ab92ea7f4d323420088dd8c9775f180cd
commit-date: 2020-05-04
host: x86_64-pc-windows-msvc
release: 1.43.1
LLVM version: 9.0
```
